Mathlib/Order/Notation.lean

```diff
@@ -54,11 +55,66 @@ class Inf (α : Type*) where
 
 attribute [ext] Min Max
 
-@[inherit_doc]
-infixl:68 " ⊔ " => Max.max
+/--
+Least upper bound (`\lub` notation). `x ⊔ y` is the same as `max x y`
+and it is preferred when the type of `x` and `y` is not a linear order.
+-/
+syntax:68 term:68 " ⊔ " term:69 : term
 
-@[inherit_doc]
-infixl:69 " ⊓ " => Min.min
+/--
+Greatest lower bound (`\glb` notation). `x ⊓ y` is the same as `min x y`
+and it is preferred when the type of `x` and `y` is not a linear order.
+-/
+syntax:69 term:69 " ⊓ " term:70 : term
+
+macro_rules
+| `($a ⊔ $b) => `(Max.max $a $b)
+| `($a ⊓ $b) => `(Min.min $a $b)
+
+section Delab
+
+open Lean Meta PrettyPrinter Delaborator SubExpr Qq
+
+/--
+Return `true` if `LinearOrder` is imported and there is an instance of `LinearOrder e`.
+
+We use a `try catch` block to make sure there are no surprising errors during delaboration.
+-/
+private def hasLinearOrder (u : Level) (e : Q(Type u)) (cls : Q(Type u → Type u))
+    (inst : Q($cls $e)) : MetaM Bool := do
+  try
+    guard (← hasConst `LinearOrder) -- avoid a type class search when `LinearOrder` isn't defined.
+    let linearOrder : Q(Type u → Type u) := .const `LinearOrder [u]
+    let ofLinearOrder ←
+      withLocalDeclDQ `α q(Type $u) fun fvarType =>
+        withLocalDeclDQ `inst q($linearOrder $fvarType) fun fvarInst => do
+          let ofLinearOrder ← synthInstanceQ q($cls $fvarType)
+          return ofLinearOrder.abstract #[fvarType, fvarInst]
+    let mvarInst ← mkFreshExprMVarQ q($linearOrder $e)
+    let ofLinearOrder : Q($cls $e) := ofLinearOrder.instantiate #[e, mvarInst]
+    isDefEq ofLinearOrder inst
+  catch _ =>
+    return false
+
+@[delab app.Max.max]
+def elabSup : Delab := do
+  let_expr f@Max.max α inst _ _ := ← getExpr | failure
+  have u := f.constLevels![0]!
+  if ← hasLinearOrder u α q(Max.{u}) inst then failure -- use the default delaborator
```

Comment:
Another approach you could use here is
```lean
  let linearOrder : Q(Type u → Type u) := .const `LinearOrder [u]
  let linearOrderToMax := Q((a : Type u) → $linearOrder a → Max a) := .const `LinearOrder.toMax [u]
```
and then you could avoid the intermediate `synthInstanceQ`